### PR TITLE
[16.07] Various bits of hardening from #4604

### PR DESCRIPTION
**Pull request by natefoo**
Backport of #4604 since there are things in there that could probably be exploited on a release that doesn't include these changes, if you looked through them hard enough.

### CODE REVIEW

**Comment by hexylena** [comment]
:+1:

File: lib/galaxy/jobs/deferred/pacific_biosciences_smrt_portal.py [hunk 1]
```diff
@@ -2,13 +2,15 @@
 Module for managing jobs in Pacific Bioscience's SMRT Portal and automatically transferring files
 produced by SMRT Portal.
 """
-import json
 import logging
 import urllib2
```
Comment (nsoranzo) [inline lib/galaxy/jobs/deferred/pacific_biosciences_smrt_portal.py]: You can remove the `urllib2` import.

File: lib/galaxy/jobs/runners/util/kill.py [hunk 2]
```diff
@@ -1,4 +1,5 @@
 import os
+import subprocess
 from platform import system
 from time import sleep
 from subprocess import Popen
```
Comment (nsoranzo) [inline lib/galaxy/jobs/runners/util/kill.py]: Remove this unused import.

**Comment by nsoranzo** [comment]
There are a 2 `subprocess.Popen()` calls that could also be updated in `lib/galaxy/datatypes/converters/sam_to_bam.py`

Also, you need to backport #4699 and https://github.com/galaxyproject/galaxy/pull/4720.

File: lib/tool_shed/capsule/capsule_manager.py [hunk 3]
```diff
@@ -9,6 +9,7 @@
 from time import gmtime
```
Comment (nsoranzo) [inline lib/tool_shed/capsule/capsule_manager.py]: Remove `urllib` import on the line above this.

File: lib/galaxy/datatypes/tabular.py [hunk 4]
```diff
@@ -515,15 +515,12 @@ def merge( split_files, output_file):
         Multiple SAM files may each have headers. Since the headers should all be the same, remove
         the headers from files 1-n, keeping them in the first file only
         """
-        cmd = 'mv %s %s' % ( split_files[0], output_file )
-        result = os.system(cmd)
-        if result != 0:
-            raise Exception('Result %s from %s' % (result, cmd))
+        shutil.move(split_files[0], output_file)
```
Comment (nsoranzo) [inline lib/galaxy/datatypes/tabular.py:519]: You need to add:
```
import shutil
```
at line 11.

**Comment by natefoo** [comment]
@nsoranzo Thanks for the review! Is #4720 directly related? It doesn't look like it. If not, I should probably open an additional PR for it.

**Comment by nsoranzo** [comment]
@natefoo #4720 is related in the sense that removing `shell=True` from the `subprocess` call in `lib/galaxy/jobs/runners/util/job_script/__init__.py` makes it faster and so exacerbates the already existing problem in `_handle_script_integrity()`, which is the wrong placement of the `time.sleep()` call, see https://github.com/galaxyproject/galaxy/pull/4720#discussion_r141585989 .

File: lib/galaxy/web/proxy/__init__.py [hunk 5]
```diff
@@ -243,20 +244,16 @@ def handle_requests(self, authentication, proxy_requests, route_name, container_
...
+            requests.get(self.api_url, headers={'Content-Type': 'application/json'}, data=json.dumps(values))
+        except requests.exceptions.ConnectionError as err:
+            log.exception(err)
             if sleep > 5:
                 excp = "Could not contact proxy after %s seconds" % sum(range(sleep + 1))
                 raise Exception(excp)
             time.sleep(sleep)
-            self.handle_requests(authentication, proxy_requests, route_name, container_ids, sleep=sleep + 1)
-        pass
+            self.handle_requests(authentication, proxy_requests, route_name, container_ids, container_interface, sleep=sleep + 1)
```
Comment (nsoranzo) [inline lib/galaxy/web/proxy/__init__.py]: `./lib/galaxy/web/proxy/__init__.py:257:93: F821 undefined name 'container_interface'`

You need to remove `container_interface, ` from this call, the parameter was added in release 17.05: https://github.com/galaxyproject/galaxy/pull/3863/files#diff-89850e7b31329e95fcf2e0475b5769d5L302

**Review by nsoranzo — APPROVED** [review]

**Comment by natefoo** [comment]
Thanks @nsoranzo !

**Pull request merged**